### PR TITLE
Merge Led and Off-season

### PR DESCRIPTION
**Pull request by PureRandomGit**

### CODE REVIEW

File: src/main/java/frc/robot/RobotContainer.java [hunk 1]
```diff
@@ -45,19 +49,28 @@ public class RobotContainer {
...
     public RobotContainer() {
         swerveSubsystem = new Swerve();
         shooterSubsystem = new Shooter();
         pivotSubsystem = new Pivot(shooterSubsystem);
         climberSubsystem = new Climber();
         intakeSubsystem = new Intake();
-        controllers = Controllers.getInstance();
+        espSubsystem = new ESP(); 
+         
+
```
Comment (brothersw) [inline src/main/java/frc/robot/RobotContainer.java:69]: who so much whitespace?

File: src/main/java/frc/robot/RobotContainer.java [hunk 2]
```diff
@@ -89,24 +103,26 @@ public RobotContainer() {
     private void configureButtonBindings() {
         /* -------- drive code -------- */
 
+        
+
```
Comment (brothersw) [inline src/main/java/frc/robot/RobotContainer.java:107]: Same with white space here

File: src/main/java/frc/robot/RobotContainer.java [hunk 3]
```diff
@@ -161,6 +177,8 @@ private boolean isBlue() {
         }
     }
 
+
```
Comment (brothersw) [inline src/main/java/frc/robot/RobotContainer.java:180]: same with white space here

File: src/main/java/frc/robot/commands/LEDCommand.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+package frc.robot.commands;
+import edu.wpi.first.wpilibj2.command.Command;
+import frc.robot.subsystems.ESP;
+import frc.robot.subsystems.Shooter;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+
+public class LEDCommand extends Command {
+  /** Creates a new DigitalCommand. */
+  ESP esp;
```
Comment (brothersw) [inline src/main/java/frc/robot/commands/LEDCommand.java:13]: I would rename this

File: src/main/java/frc/robot/RobotContainer.java [hunk 5]
```diff
@@ -45,19 +49,28 @@ public class RobotContainer {
     Controllers controllers;
     public AutoDriveBuilder autoDriveBuilder;
     public AutoShootBuilder autoShootBuilder;
-
+    public ESP espSubsystem;
```
Comment (brothersw) [inline src/main/java/frc/robot/RobotContainer.java:52]: I would rename this

File: src/main/java/frc/robot/commands/LEDCommand.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  // Called when the command is initially scheduled.
+  @Override
+  public void initialize() {
+    esp.setLED(true); 
+  }
+
+  // Called every time the scheduler runs while the command is scheduled.
+  @Override
+  public void execute() {
+    SmartDashboard.putBoolean("topSensor", shooter.topPieceSensorActive());
```
Comment (brothersw) [inline src/main/java/frc/robot/commands/LEDCommand.java:32]: this shouldn't be where smart dashboard puts data, that should be from the subsystem and I believe it already did that

File: src/main/java/frc/robot/subsystems/ESP.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+import frc.robot.Constants;
+
+public class ESP extends SubsystemBase {
+  /** Creates a new ESP. */
+  DigitalOutput output; 
+  
+  public ESP() {
+    output = new DigitalOutput(Constants.LED_PORT); 
+  }
+
```
Comment (brothersw) [inline src/main/java/frc/robot/subsystems/ESP.java:19]: please add some javadocs

File: src/main/java/frc/robot/subsystems/ESP.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+// the WPILib BSD license file in the root directory of this project.
+
+package frc.robot.subsystems;
+
+import edu.wpi.first.wpilibj.DigitalOutput;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+
+public class ESP extends SubsystemBase {
```
Comment (brothersw) [inline src/main/java/frc/robot/subsystems/ESP.java:12]: I would rename this

**Review by brothersw — CHANGES_REQUESTED** [review]
looks great! just some minor tweaks and it should be ready

File: src/main/java/frc/robot/RobotContainer.java (hunk 2 above)
Comment (PureRandomGit) [reply src/main/java/frc/robot/RobotContainer.java]: Removed

File: src/main/java/frc/robot/commands/LEDCommand.java (hunk 4 above)
Comment (PureRandomGit) [reply src/main/java/frc/robot/commands/LEDCommand.java]: Renamed to ESPController

File: src/main/java/frc/robot/commands/LEDCommand.java (hunk 6 above)
Comment (PureRandomGit) [reply src/main/java/frc/robot/commands/LEDCommand.java]: Removed

File: src/main/java/frc/robot/subsystems/ESP.java (hunk 7 above)
Comment (PureRandomGit) [reply src/main/java/frc/robot/subsystems/ESP.java]: Added, Please review to ensure correct format.

File: src/main/java/frc/robot/RobotContainer.java (hunk 3 above)
Comment (PureRandomGit) [reply src/main/java/frc/robot/RobotContainer.java]: removed

File: src/main/java/frc/robot/subsystems/ESPController.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+// the WPILib BSD license file in the root directory of this project.
+
+package frc.robot.subsystems;
+
+import edu.wpi.first.wpilibj.DigitalOutput;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+
+public class ESPController extends SubsystemBase {
```
Comment (brothersw) [inline src/main/java/frc/robot/subsystems/ESPController.java:12]: ESP Controller doesn't describe a system that controls leds but some generic ESP32

**Review by brothersw — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/ESPController.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  /** Creates a new ESP controller. */
+  DigitalOutput output; 
+  
+  /**
+   * Controlls an ESP32 via digital pin
+   * @param LED_PORT The ESP32's digital pin
```
Comment (brothersw) [inline src/main/java/frc/robot/subsystems/ESPController.java:18]: there are no parameters here

File: src/main/java/frc/robot/subsystems/ESPController.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+   */
+  public ESPController() {
+    output = new DigitalOutput(Constants.LED_PORT); 
+  }
+
+  public void setLED(boolean isGreen){
```
Comment (brothersw) [inline src/main/java/frc/robot/subsystems/ESPController.java:24]: this doesn't have a javadoc

**Review by brothersw — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/ESPController.java (hunk 9 above)
Comment (PureRandomGit) [reply src/main/java/frc/robot/subsystems/ESPController.java]: Renamed to LED

File: src/main/java/frc/robot/subsystems/ESPController.java (hunk 10 above)
Comment (PureRandomGit) [reply src/main/java/frc/robot/subsystems/ESPController.java]: Removed parameter

File: src/main/java/frc/robot/subsystems/ESPController.java (hunk 11 above)
Comment (PureRandomGit) [reply src/main/java/frc/robot/subsystems/ESPController.java]: Added Javadoc

File: src/main/java/frc/robot/commands/LEDCommand.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+package frc.robot.commands;
+import edu.wpi.first.wpilibj2.command.Command;
+import frc.robot.subsystems.LED;
+import frc.robot.subsystems.Shooter;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+
+public class LEDCommand extends Command {
+  /** Creates a new DigitalCommand. */
+  LED esp;
```
Comment (brothersw) [inline src/main/java/frc/robot/commands/LEDCommand.java:13]: The variable name is still called esp even when the subsystem name changed

**Review by brothersw — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/RobotContainer.java [hunk 13]
```diff
@@ -45,19 +47,25 @@ public class RobotContainer {
     Controllers controllers;
     public AutoDriveBuilder autoDriveBuilder;
     public AutoShootBuilder autoShootBuilder;
-
+    public LED espSubsystem;
```
Comment (brothersw) [inline src/main/java/frc/robot/RobotContainer.java:50]: it is still called esp subsystem here

File: src/main/java/frc/robot/RobotContainer.java [hunk 14]
```diff
@@ -45,19 +47,25 @@ public class RobotContainer {
...
     public AutoDriveBuilder autoDriveBuilder;
     public AutoShootBuilder autoShootBuilder;
-
+    public LED espSubsystem; 
     Command floorIntakeCommand;
     Command outtakeCommand;
     Command hpIntakeCommand;
+    Command ledCommand; 
+    Command testCommand; 
+    Command analogCommand;
```
Comment (brothersw) [inline src/main/java/frc/robot/RobotContainer.java:56]: also, you have 2 unused commands here

File: src/main/java/frc/robot/RobotContainer.java [hunk 15]
```diff
@@ -45,19 +47,25 @@ public class RobotContainer {
...
     public RobotContainer() {
         swerveSubsystem = new Swerve();
         shooterSubsystem = new Shooter();
         pivotSubsystem = new Pivot(shooterSubsystem);
         climberSubsystem = new Climber();
         intakeSubsystem = new Intake();
+        espSubsystem = new LED(); 
         controllers = Controllers.getInstance();
-
+        ledCommand = new LEDCommand(shooterSubsystem, espSubsystem);
```
Comment (brothersw) [inline src/main/java/frc/robot/RobotContainer.java:67]: to reduce complexity of stuff stored everywhere, it is probably better to instantiate it directly into setDefaultCommand

**Review by brothersw — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/RobotContainer.java (hunk 13 above)
Comment (PureRandomGit) [reply src/main/java/frc/robot/RobotContainer.java:50]: fixed

File: src/main/java/frc/robot/RobotContainer.java (hunk 14 above)
Comment (PureRandomGit) [reply src/main/java/frc/robot/RobotContainer.java:56]: Removed

File: src/main/java/frc/robot/RobotContainer.java (hunk 15 above)
Comment (PureRandomGit) [reply src/main/java/frc/robot/RobotContainer.java]: Changed

**Review by brothersw — APPROVED** [review]
lgtm!

**Pull request merged**